DataDog/dd-trace-java
# Create S3 instrumentation + add span pointers

# What Does This Do
Creates an instrumentation for S3 and adds span pointers in S3 for `putObject`, `copyObject`, and `completeMultipartUpload` requests.

Span pointers are similar to [Span Links](https://docs.datadoghq.com/tracing/trace_collection/span_links/), but for cases when it is impossible to pass the Trace ID and Span ID between the spans that need to be linked.

When the calculated hashes for the upstream and downstream lambdas match, the Datadog backend will automatically link the two traces together.

<img width="1046" alt="Screenshot 2024-12-10 at 2 17 55 PM" src="https://github.com/user-attachments/assets/d75205d9-d682-45f3-9127-c1375479f687">
<img width="1040" alt="Screenshot 2024-12-10 at 2 18 10 PM" src="https://github.com/user-attachments/assets/1e15ec90-20e3-4937-a081-d1dbe9cdd689">
<img width="1044" alt="Screenshot 2024-12-10 at 2 18 29 PM" src="https://github.com/user-attachments/assets/c31614bf-79bc-4fa1-811b-1416efbb9135">
<img width="1042" alt="Screenshot 2024-12-10 at 2 31 33 PM" src="https://github.com/user-attachments/assets/f9b8b4b5-6b1d-49cf-984c-5aa773e792d8">

When clicking on the linked span, a new tab opens linking to the downstream Lambda function that was triggered by this S3 object update.

This feature can be disabled by setting the config `trace.aws.add.span.pointers` or Lambda environment variable `TRACE_AWS_ADD_SPAN_POINTERS` to false.

# Motivation

Span pointers is a new feature being developed by the Serverless team. This feature already exists in Python & Node, and I'm working on adding it to other runtimes (.NET, Java, Golang).

# Testing Guidelines

Easy: Checkout [this span](https://ddserverless.datadoghq.com/functions?cloud=aws&config_serverless-auto-span-links=true&config_trace_show_hidden_metadata=true&entity_view=lambda_functions&env=dev&fromUser=false&graphType=flamegraph&panel_end=1733859287404&panel_paused=false&panel_start=1733858387404&shouldShowLegend=true&sort=time&sp=%5B%7B%22p%22%3A%7B%22entityId%22%3A%22aws-lambda-functions%2Bnhulston-java-s3-span-pointers-dev-main%2Bus-east-1%2B425362996713%22%7D%2C%22i%22%3A%22lambda-panel%22%7D%2C%7B%22p%22%3A%7B%22traceID%22%3A%222542693623706377411%22%2C%22selectedSpanID%22%3A%222302740288124301958%22%7D%2C%22i%22%3A%22trace-panel%22%7D%5D&spanID=2302740288124301958&spanViewType=span-links&text_search=nhulston&traceID=2542693623706377411&start=1733857860000&end=1733858760000&paused=false), which is an upstream Lambda that makes all three S3 requests and triggers 3 separate downstream Lambda invocations. Therefore, the 3 span pointers should link to 3 different traces. Each downstream trace should point back to this span.
(Note: this trace might be from so long ago that the linked spans are no longer found, but just check that the span links themselves are created)

More thorough: Run [this Lambda function](https://us-east-1.console.aws.amazon.com/lambda/home?region=us-east-1#/functions/nhulston-java-s3-span-pointers-dev-main?tab=testing) (only on the serverless AWS account) with the event payload
```json
{
  "shouldPutObject": false,
  "shouldCopyObject": false,
  "shouldMultipartUpload": false,
}
```
and change one of the bools to true. Check Datadog to ensure that the spans are properly linked.

I also added tests:
```shell
# Unit tests
./gradlew :dd-trace-core:test --tests "*.SpanPointersProcessorTest"

# Integration tests
./gradlew :dd-java-agent:instrumentation:aws-java-s3-2.0:test
```

# Additional Notes
- Changes are organized in commits (so you have an easier time reviewing)
- This handles the 'upstream' case, for Lambdas that make S3 requests. I will have another PR in both Lambda agents to handle the 'downstream' case, for Lambdas that are triggered by S3 updates.
- S3 and Dynamo are supported in Python+Node. I will soon open another PR in this tracer to add support for Dynamo.

# Contributor Checklist

- [x] Format the title [according the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- [x] Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any usefull labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- [ ] Squash your commits prior merging or merge using GitHub's [Squash and merge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits)
- [x] Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- [ ] Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [SVLS-6060]



[SVLS-6060]: https://datadoghq.atlassian.net/browse/SVLS-6060?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
